Services/Concurrency/RatedCounter.cs

```diff
@@ -0,0 +1,160 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Diagnostics;
+using Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Exceptions;
+
+namespace Microsoft.Azure.IoTSolutions.DeviceSimulation.Services.Concurrency
+{
+    public class PerSecondCounter : RatedCounter
+    {
+        public PerSecondCounter(double rate, string name, ILogger logger)
+            : base(rate, 1000, name, logger)
+        {
+        }
+    }
+
+    public class PerMinuteCounter : RatedCounter
+    {
+        public PerMinuteCounter(double rate, string name, ILogger logger)
+            : base(rate, 60 * 1000, name, logger)
+        {
+        }
+    }
+
+    // TODO: optimize the memory usage for this counter (see Queue<long> usage)
+    public class PerDayCounter : RatedCounter
+    {
+        public PerDayCounter(double rate, string name, ILogger logger)
+            : base(rate, 86400 * 1000, name, logger)
+        {
+        }
+    }
+
+    // Leaky bucket counter
+    public abstract class RatedCounter
+    {
+        // At least 1 second duration
+        private const double MIN_TIME_UNIT = 1000;
```

Comment:
_MS?

Reply:
I'd have to add "ms" or "msecs" to all the vars, which I did at some point and made the code harder to read. I prefer the assumption that all the timing values are expressed in msecs by definition, see also the comment on line 46